### PR TITLE
406 support enabling/disabling particular atoms

### LINKED ISSUE #406: Support enabling/disabling particular atoms
## Details

- Related to #405

## Subtasks

- [x] Add `normalizer` options to disable/enable particular atoms.
- [x] Add `normalizer` commands to print arguments for the pipeline script.
- [x] Update pipeline config to use normalizer command that prints arguments.
- [ ] Update docs about normalizer options.
- [x] Update docs about pipeline config.

### PR DESCRIPTION
**Pull request by deemp**
## Issues
-  #406
- https://github.com/objectionary/normalizer/issues/431

---

## PR-Codex overview
This PR updates the `eo-phi-normalizer` package by adding new phi files, updating pipeline configurations, and introducing new rules.

### Detailed summary
- Added new phi files for `error`, `dataized`, `cti`, `try`, `rust`, `string`, `cage`, `true`, `false`, `nan`, `go`, and `while`
- Updated pipeline configurations and hashes
- Modified `TH.hs`, `Config.hs`, `RunYegor.hs`, `Fast.hs`
- Introduced new rules and functions in various files

> The following files were skipped due to too many changes: `eo-phi-normalizer/data/0.38.4/org/eolang/while.phi`, `eo-phi-normalizer/data/0.38.4/org/eolang/seq.phi`, `eo-phi-normalizer/data/0.38.4/org/eolang/int.phi`, `eo-phi-normalizer/data/0.38.4/org/eolang/switch.phi`, `eo-phi-normalizer/src/Language/EO/Phi/Pipeline/Dataize/PrintConfigs.hs`, `eo-phi-normalizer/data/0.38.4/org/eolang/malloc.phi`, `eo-phi-normalizer/data/0.38.4/org/eolang/bytes.phi`, `eo-phi-normalizer/src/Language/EO/Phi/Rules/Common.hs`, `eo-phi-normalizer/data/0.38.4/org/eolang/tuple.phi`, `eo-phi-normalizer/data/0.38.4/org/eolang/float.phi`, `site/docs/src/pipeline.md`, `scripts/pipeline.sh`, `eo-phi-normalizer/data/0.38.4/org/eolang/positive-infinity.phi`, `eo-phi-normalizer/data/0.38.4/org/eolang/negative-infinity.phi`, `eo-phi-normalizer/eo-phi-normalizer.cabal`, `eo-phi-normalizer/app/Main.hs`, `eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs [hunk 1]
```diff
@@ -184,27 +184,33 @@ evaluateBinaryDataizationFunChain ::
...
-      return resultObj
-    (Left _l, Left _r) ->
-      fail (name <> ": Couldn't find bytes in both LHS and RHS")
-    (Left l, _) -> do
-      fail (name <> ": Couldn't find bytes in LHS: " <> printTree (hideRho l))
-    (_, Left r) -> do
-      fail (name <> ": Couldn't find bytes in RHS: " <> printTree (hideRho r))
-  return (result, ())
+  ctx <- getContext
+  if HashSet.member (DisabledAtomName name) ctx.disabledAtomNames
```
Comment (fizruk) [inline eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs:188]: This is not the correct place for this check. Instead, use the check in the `dataizeStepChain` function, before `evaluateBuiltinFunChain` is called.

**Comment by 0crat** [comment]
@fizruk Thanks for the review! You've earned +25 points for this.

**Comment by deemp** [comment]
To enable or disable particular atoms for a test set, add an `atoms` object into the test set in the pipeline config (the file path is `pipeline/config.yaml`).

Here's an example of a test set where the `foo` atom is enabled and the `bar` atom is disabled.

```yaml
- eo:
    original: eo/eo-runtime/src/test/eo/org/eolang/as-phi-tests.eo
    yaml: pipeline/eo-yaml/as-phi-tests.yaml
    filtered: pipeline/eo-filtered/as-phi-tests.eo
  phi:
    initial: pipeline/phi-initial/as-phi-tests.phi
    normalized: pipeline/phi-normalized/as-phi-tests.phi
    bindings-path-initial: org.eolang
    bindings-path-normalized: org.eolang
  atoms:
    enable:
      - foo
    disable:
      - bar
```

**Review by fizruk — APPROVED** [review]

File: eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs [hunk 2]
```diff
@@ -50,14 +52,20 @@ dataizeStepChain obj@(Formation bs)
...
-      logStep ("Evaluating lambda '" <> funcName <> "'") (Left obj)
-      msplit (evaluateBuiltinFunChain funcName obj ()) >>= \case
-        Nothing -> do
-          ctx <- getContext
-          return (ctx, Left obj)
-        Just ((obj', _state), _alts) -> do
-          ctx <- getContext
-          return (ctx, Left obj')
+      ctx' <- getContext
+      if HashSet.member funcName knownAtomNames && not (HashSet.member funcName ctx'.enabledAtomNames)
```
Comment (fizruk) [inline eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs:56]: I would factor this condition into a separate function.

File: eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs [hunk 3]
```diff
@@ -270,87 +278,129 @@ evaluateBytesBytesFunChain name = evaluateUnaryDataizationFunChain name intToByt
 evaluateFloatFloatFloatFunChain :: String -> (Double -> Double -> Double) -> Object -> EvaluationState -> DataizeChain (Object, EvaluationState)
 evaluateFloatFloatFloatFunChain name = evaluateBinaryDataizationFunChain name floatToBytes bytesToFloat wrapBytesInFloat extractRho (extractLabel "x")
 
+knownAtoms :: [(String, String -> Object -> EvaluationState -> DataizeChain (Object, EvaluationState))]
+knownAtoms =
+  [ ("Lorg_eolang_int_gt", \name obj -> evaluateIntIntBoolFunChain name (>) obj)
```
Comment (fizruk) [inline eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs:283]: If you change the order of arguments, you can avoid lambdas and simply write `evaluateIntIntBoolFunChain (>)`.

File: eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs [hunk 4]
```diff
@@ -270,87 +278,129 @@ evaluateBytesBytesFunChain name = evaluateUnaryDataizationFunChain name intToByt
...
-evaluateBuiltinFunChain atomName obj = \state -> do
-  logStep ("[WARNING]: unknown atom (" <> atomName <> ")") (Left obj)
+evaluateBuiltinFunChain name obj =
+  case HashMap.lookup name knownAtomsSet of
+    Just f -> f name obj
+    Nothing -> evaluateBuiltinFunChainUnknown name obj
+
+evaluateBuiltinFunChainUnknown :: String -> a -> b1 -> Chain (Either a b2) (a, b1)
+evaluateBuiltinFunChainUnknown atomName obj state = do
+  logStep [fmt|[WARNING]: unknown atom ({atomName})|] (Left obj)
```
Comment (fizruk) [inline eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs:403]: Unknown atoms are normal and should not be warnings.

File: eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs (hunk 3 above)
Comment (deemp) [reply eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs]: Addressed in https://github.com/objectionary/normalizer/pull/426/commits/32be6c98f22e44833723d70e3c27f0607e56cdd5

File: eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs (hunk 4 above)
Comment (deemp) [reply eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs]: Addressed in https://github.com/objectionary/normalizer/pull/426/commits/75fb7691c7e8842a7ae673d77dcccd8142d9bb28

File: eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs (hunk 2 above)
Comment (deemp) [reply eo-phi-normalizer/src/Language/EO/Phi/Dataize.hs]: Addressed in https://github.com/objectionary/normalizer/pull/426/commits/32be6c98f22e44833723d70e3c27f0607e56cdd5

However, I'm not sure I understood your suggestion correctly.

**Review by fizruk — APPROVED** [review]

**Comment by 0crat** [comment]
@deemp Thanks for the contribution! You've earned +5 points for this: +20 as a basis; +5 for 2236 hits-of-code; -7 for too many hits-of-code (2236 >= 100); -15 for way too many hits-of-code (2236 >= 400); -9 for 9 comments; +11 to give you at least something. Please, [keep](https://www.yegor256.com/2018/03/06/speed-vs-quality.html) them coming.

**Pull request merged**